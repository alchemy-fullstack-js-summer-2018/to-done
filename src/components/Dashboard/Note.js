import React, { Component } from 'react';
import NoteDisplay from './NoteDisplay';
import NoteForm from './NoteForm';
import PropTypes from 'prop-types';

class Note extends Component {

  state = {
    editing: false
  };

  static propTypes = {
    note: PropTypes.object.isRequired,
  };

  handleEndEdit = () => {
    this.setState({ editing: false });
  };

  render() {
    const { editing } = this.state;
    const { note } = this.props;

    return (
      <li>
        {editing
          ? <NoteForm 
            note={note}
            onCancel={this.handleEndEdit}
          />
          : <NoteDisplay 
            note={note}
          />
        }
      </li>
    );
  }
}

export default Note;