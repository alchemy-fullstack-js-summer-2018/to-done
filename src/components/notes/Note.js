import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NoteDisplay from './NoteDisplay';
import NoteForm from './NoteForm';

class Note extends Component {
  state = {
    editing: false
  };

  static propTypes = {
    note: PropTypes.object.isRequired,
    onUpdate: PropTypes.func.isRequired,
    onRemove: PropTypes.func.isRequired
  };

  handleEdit = () => {
    this.setState({ editing: true });
  };

  handleDelete = () => {
    const { note, onRemove } = this.props;
    return onRemove(note.key);
  };

  handleComplete = note => {
    const { onUpdate } = this.props;
    return onUpdate(note).then(this.handleEndEdit);
  };

  handleEndEdit = () => {
    this.setState({ editing: false });
  };

  render() { 
    const { editing } = this.state;
    const { note } = this.props;

    return (
      <li>
        {editing
          ? <NoteForm
            note={note}
            onComplete={this.handleComplete}
            onCancel={this.handleEndEdit}
          />
          : <NoteDisplay
            note={note}
            onEdit={this.handleEdit}
            onDelete={this.handleDelete}
          />
        }
      </li>
    );
  }
}
 
export default Note;