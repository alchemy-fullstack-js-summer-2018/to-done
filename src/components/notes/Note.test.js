import Note from './Note';
import React from 'react';
import { shallow } from 'enzyme';

describe('Note', () => {
  it('renders display or edit', () => {
    const handleRemove = jest.fn();
    const handleUpdate = jest.fn();

    const note = { key: 'abc', title: 'hi', content: 'hello', completed: false };

    const wrapper = shallow(<Note
      note={note}
      onRemove={handleRemove}
      onUpdate={handleUpdate}
    />);

    const component = wrapper.instance();

    expect(wrapper.state('editing')).toBe(false);
    component.handleEdit();
    expect(wrapper.state('editing')).toBe(true);
    component.handleEndEdit();
    expect(wrapper.state('editing')).toBe(false);

    component.handleDelete();

    const removeCalls = handleRemove.mock.calls;
    expect(removeCalls.length).toBe(1);
    expect(removeCalls[0][0]).toEqual(note.key);
  });
});