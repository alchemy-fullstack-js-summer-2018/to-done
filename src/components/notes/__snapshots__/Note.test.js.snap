// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note Renders display or edit 1`] = `
<li>
  <NoteDisplay
    note={
      Object {
        "content": "Portland is sunny today",
        "key": "abc",
        "title": "Note #1",
      }
    }
    onDelete={[Function]}
    onEdit={[Function]}
  />
</li>
`;
